Support github actions with Windows 10 SDK and VS 2022 (v142)

ader1990 opened the pull request:
This change introduces support to build the Cloudbase-init MSI with recent versions of Visual Studio, required for Github Actions. At the same time, introduces a Github Actions workflow to build, install and functionally test the MSI.

From an implementation point of view - these are the changes:

  * UtilsActions/UtilsActions.vcxproj -> target v142 version (required for new VS versions)
  * Adds CloudbaseInitSetup/Binaries/x64/UtilsActions.dll binaries in case we retire the UtilsActions project
  * Remove VC90 msms (not needed)
  * Remove VC140Redist (not needed)
  * Fix wix bug related to the Heat directory (required for new VS versions)
  * Use a relative python folder path instead of hardcoded one (required for github actions)
  * SetVCVars now can leverage an Automatic detection of what Visual Studio is installed (required for github actions running on multiple Windows versions in case we want to change the runner from the current Windows 2022 to Windows 2025)
  * Adds github action to build and functionally test the MSI (install MSI, run cloudbase-init, validate the run)

ader1990 committed:
binaries: add UtilsActions.dll

Visual Studio 2022 or later cannot build the current version of
UtilsActions project, because of the type errors.

Having these UtilsActions.dll binaries is actually beneficial as there
is no more need of the UtilsActions project, at the expense of not being
able to modify those methods if needed.

But the UtilsActions.dll methods are rather fixed in their purpose and
should not change for the foreseeable future.

If, for any reason, there is the need to change those methods, the
project will need to be rewritten. Currently, the rewrite is not needed,
thus this addition.

```
std::wstring GenerateRandomPassword();
void GetUserSid(const std::wstring& username, PSID &pSid);
std::vector<std::wstring> GetUserRights(const std::wstring& username);
void AssignUserRights(const std::wstring& username, const
std::vector<std::wstring> rights);
```

Signed-off-by: Adrian Vladu <avladu@cloudbasesolutions.com>

ader1990 committed:
github_workflows: add boilerplate

ader1990 committed:
github_workflows: add vs code 2022 and relative path options

ader1990 committed:
add redist msms

Signed-off-by: Adrian Vladu <avladu@cloudbasesolutions.com>

ader1990 committed:
remove utils actions project

ader1990 committed:
Fix wix error on VS 2022

ader1990 committed:
Fix relative root paths

ader1990 committed:
Revert "remove utils actions project"

This reverts commit 88b19384d25d6818a2eed2b9ccef6f5c565f7d21.

ader1990 committed:
UtilsActions: bump platformtoolset from 141 to 143

ader1990 committed:
UtilsActions: import string

Signed-off-by: Adrian Vladu <avladu@cloudbasesolutions.com>

petrutlucian94 commented on .github/workflows/build_cbsinit.yml:
```diff
@@ -0,0 +1,36 @@
+name: Cloudbase-Init build MSI retest
```
nit: I'd drop the "retest" word, it's a bit confusing

ader1990 replied:
will do, leftover from testing.

petrutlucian94 commented on CloudbaseInitSetup/CloudbaseInitSetup.wixproj:
```diff
@@ -309,10 +309,9 @@
   <Import Project="$(WixTargetsPath)" />
   <Target Name="BeforeBuild">
-    <HeatDirectory DirectoryRefId="INSTALLDIR" OutputFile="Python_tmp.wxs" Directory="C:\Python_CloudbaseInit" ComponentGroupName="PythonComponentGroup" ToolPath="$(WixToolPath)" PreprocessorVariable="var.PythonSourcePath" GenerateGuidsNow="true" SuppressCom="true" SuppressRegistry="true" KeepEmptyDirectories="true" />
+    <HeatDirectory DirectoryRefId="INSTALLDIR" OutputFile="Python_tmp.wxs" Directory="Python_CloudbaseInit" ComponentGroupName="PythonComponentGroup" ToolPath="$(WixToolPath)" PreprocessorVariable="var.PythonSourcePath" GenerateGuidsNow="true" SuppressCom="true" SuppressRegistry="true" KeepEmptyDirectories="true" RunAsSeparateProcess="true" />
```
Just curious, why do we need this new ``RunAsSeparateProcess`` flag?

ader1990 replied:
https://github.com/cloudbase/cloudbase-init-installer/pull/50/commits/f128f1cfbfcf91cbd685293ab56452685fd9abe7 -> fixes a wix known issue https://github.com/wixtoolset/issues/issues/2467#issuecomment-736519622. Will add a proper commit message, there was one commit description that got lost between branches.

petrutlucian94 commented on CloudbaseInitSetup/Binaries/x86/UtilsActions.dll:
Since you're fixing the `utils` compilation, why do we need to include the DLLs?

ader1990 replied:
The msms were added to keep backwards compatibility in the sense of codebase, but I need to check if they are still required and for what version of VS and SDK. For example, if we want to only support VS 2022 or VS 2017 and 2019 and 2022, I need to verify more.

ader1990 committed:
redist: remove vc redist msms

ader1990 committed:
github_workflows: add functional test

ader1990 committed:
github_workflows: add functional test fix

ader1990 committed:
github_workflows: add functional tests on 2k22 and 2k19

Signed-off-by: Adrian Vladu <avladu@cloudbasesolutions.com>

ader1990 committed:
github_workflows: add os matrix

ader1990 committed:
github_workflows: detect version matrix

ader1990 committed:
use v142 in order to work on VS2019 github workers

ader1990 committed:
Add artifact OS suffix for no name conflicts

Signed-off-by: Adrian Vladu <avladu@cloudbasesolutions.com>

ader1990 committed:
github_workflows: use 3.13.13

Signed-off-by: Adrian Vladu <avladu@cloudbasesolutions.com>

ader1990 committed:
github_workflows: use only win 2022 runner

Windows 2019 runners do not exist anymore.

Windows 2025 has been removed as it is not needed yet.

Signed-off-by: Adrian Vladu <avladu@cloudbasesolutions.com>

ader1990 committed:
Always use relative dir build path

Signed-off-by: Adrian Vladu <avladu@cloudbasesolutions.com>

petrutlucian94 commented on .github/workflows/build_test_cbsinit.yml:
```diff
@@ -0,0 +1,101 @@
+name: Cloudbase-Init build MSI retest
```
So it's not just building the msi but also testing it. We could have separate jobs, one for building it and one for testing it, reusing the job results.

I guess it's reasonable to have a single job for now, but let's rename it to "Build and test Cloudbase-init MSI"

petrutlucian94 reviewed (approved):
Looks good overall.

ader1990 committed:
github_workflows: rename action and steps

Signed-off-by: Adrian Vladu <avladu@cloudbasesolutions.com>
